Resolving Issue 701: changing for-target-subscript to invalid-for-target

ckatsube opened the pull request:
<!--- Provide a summary of your changes in the Pull Request Title above. -->
<!--- If this is a work in progress (not yet ready to be merged), make this a draft pull request. -->

## Motivation and Context

<!--- Why is this pull request required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here: -->
<!--- https://docs.github.com/en/github/managing-your-work-on-github/managing-your-work-with-issues-and-pull-requests/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->
resolves #701

## Your Changes

<!--- Describe your changes here. -->
**Description**: changed the custom checker `InvalidForTargetChecker` (previously `ForTargetSubscriptChecker`) located in `python_ta/checkers/invalid_for_checker.py` (previously `for_target_subscript_checker.py`). This change entails updating the error message from `for-target-subscript` to `invalid-for-target` as well as changing the checker to add the message in the additional condition when a target of a for loop is an object's attribute (`for x.attr in ...`)

**Type of change** (select all that apply):
<!--- Put an `x` in all the boxes that apply. -->
<!--- Remove any lines that do not apply. --> 

- [x] New feature (non-breaking change which adds functionality)

## Testing

<!--- Please describe in detail how you tested this pull request. -->
Manually ran pylint on a test file using the CLI command `pylint --load-plugins="python_ta.checker.invalid_for_target_checker" sandbox.py` (where sandbox.py is an untracked local module containing the below code) and python_ta's `check_all` on the same test file.
<!--- This can include tests you added and manual testing. -->
```python
for x.attr in a:
    pass

for x[0] in a:
    pass
```

## Questions and Comments (if applicable)

<!-- Ask any questions you have for the maintainers of this project regarding this PR. -->
<!-- Please describe the steps you have already taken to find the answer to your question. -->
<!-- This will ensure that we can give you clear and relevant advice. -->
<!-- If you have additional comments add them here as well. -->

There are a number of other references to the `for_target_subscript` checker. I was unsure of how many of them I should update the reference to in this pull request as the checker is fully functioning (these now-outdated references appear to be related to documentation and testing.
There are also a few references to `for-target-subscript`.
![image](https://user-images.githubusercontent.com/13248104/120947959-1115a480-c70f-11eb-916b-5c05ea1aec58.png)
![image](https://user-images.githubusercontent.com/13248104/120948076-605bd500-c70f-11eb-91f6-fb8b2c1404af.png)
(Screenshot taken from PyCharm's "Find in Files" feature using the above mentioned search terms).


## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the TravisCI tests have passed. <!-- (check after opening pull request) -->
- [x] I have reviewed the test coverage changes reported on Coveralls. <!-- (check after opening pull request) -->

ckatsube commented:
@david-yz-liu 
I'm unsure of whether I should mark the PR as ready for review since I still have a few open questions. Would it be possible to get a brief review over the questions section to see if that's something I should continue working on before marking the PR as ready?

ckatsube commented:
Just remembered another thing. When looking at `invalid-for-target`, I noticed that there is a check for if the for loop's target is a `tuple`, and to check if any of the assignments in said tuple are invalid targets. You can also explicitly write a list as a for loop target in something that would look like
```python
for [a, b] in l:
    pass
```
Should I add a check for lists into this PR as well? Or create a separate PR for it.

david-yz-liu reviewed (commented):
@ckatsube thanks for the ping. Regarding your question, yes you should update *all* references to this checker, including in documentation and tests. In general it's common when updating code to update documentation and tests as well. In fact, you should create updated test cases for the changes you've made.

Also, yes please add a case for checking a `List` node as the for loop target (and tests as well). In fact, now that you pointed this out I'm thinking about cases where the for loop target is a nested expression with an invalid target in a subexpression, like

```python
for x, [y, z.attr] in lst:
```

So why don't you work on actually generalizing the checker by pulling out the actual logic that checks the target into a *recursive helper method*.

ckatsube commented on python_ta/checkers/invalid_for_target_checker.py:
```diff
@@ -6,32 +6,40 @@
+    target_is_subscript = isinstance(node.target, astroid.Subscript)
+    target_is_attr_assign = isinstance(node.target, astroid.AssignAttr)
+
+    return target_is_subscript or target_is_attr_assign
```
Noted! I'll get the above items completed before marking the PR as ready for review.

ckatsube commented:
Brief Changelog:
- Generalized `InvalidForTargetChecker` in `invalid_for_target.py` to check for target lists as well as nested tuples/lists.
- Added `test_invalid_target_checker.py` file which contains a brief test suite for the `InvalidForTargetChecker`
- Updated references of the old message `for-target-subscript` to the new `invalid-for-target`

ckatsube commented:
Unmarked as Ready for Review.
I realized that you can just use `.nodes_of_class` to extract all `Subscript` and `AssignAttr` nodes within the `For` node's target. Should we use this and check if these two lists are empty rather than flattening the target list and iterating through it?

ckatsube commented:
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

ckatsube commented:
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

ckatsube commented:
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

ckatsube commented:
Changelog:
Pretty much what the commit says: removed the helper that flattened the for node's target and instead used the `.nodes_of_class` node method to extract the invalid nodes from the target.

david-yz-liu commented on python_ta/checkers/invalid_for_target_checker.py:
```diff
@@ -0,0 +1,37 @@
+            }
+    # this is important so that your checker is executed before others
+    priority = -1
+    # pass in message symbol as a parameter of check_messages
```
You can delete this comment

david-yz-liu commented on python_ta/checkers/invalid_for_target_checker.py:
```diff
@@ -0,0 +1,37 @@
+
+    @check_messages('invalid-for-target')
+    def visit_for(self, node: astroid.For) -> None:
+
```
Delete this line

david-yz-liu reviewed (commented):
@ckatsube great, I left a few minor comments that you should be able to resolve pretty quickly!
